Remove the custom binding for ReadableStream cancel

youennf opened the pull request:
#### 2197c9117f451116809cf36bec7b9acdb0babb5e
<pre>
Remove the custom binding for ReadableStream cancel
<a href="https://rdar.apple.com/161897092">rdar://161897092</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=300107">https://bugs.webkit.org/show_bug.cgi?id=300107</a>

Reviewed by Chris Dumez.

We do a refactoring to prepare ground for Readable byte stream native sources (especially fetch responses).
This patch does the following:
- Remove custom binding for ReadableStream and reader cancel methods. We update the binding generator so that methods can return their own promise. This is necessary to support the byte stream code path as well as the default stream code path (JS built-in).
- We do a small refactoring to simplify the construction of a ReadableStreamDefaultReader so that it always has a ReadableStream. This is now necessary for the cancel code path.
- We update getReader accordingly, which gets closer to the step-by-step spec.
- We make ReadableStream a ContextDestructionObserver so that we can get its JSDOMGlobalObject, which is used when cancelling with a DOMException (used by native sources).

We also update the conversion of DOMPromise to JSValue by using guardedObject() directly.
This removes hitting the ASSERT in DOMPromise::promise(), which is similar to the the regular JSDOMPromiseDeferred.h callPromiseFunction code path.

Covered by existing tests.

Canonical link: <a href="https://commits.webkit.org/301120@main">https://commits.webkit.org/301120@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/11991aa02f0a66dc418310db6f9515b1656d0f17

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124876 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44547 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35283 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131722 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76776 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ad808ba1-f769-465f-9780-34f4b191020b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126753 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45248 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53115 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95030 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63055 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/407c5d66-9703-4547-b188-4c31e9a2f8c6) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127830 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36102 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111681 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75587 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2aad058a-6f95-49b1-8277-266ce8540802) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35037 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29838 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75199 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105862 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30066 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134390 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51717 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39520 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103508 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52134 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107896 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103272 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26322 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48629 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26920 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48738 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51593 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57401 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50982 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54338 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52675 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c31b5d77f1e72827f81c2df114f526dc930597c7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51755

youennf committed:
Remove the custom binding for ReadableStream cancel
rdar://161897092
https://bugs.webkit.org/show_bug.cgi?id=300107

Reviewed by Chris Dumez.

We do a refactoring to prepare ground for Readable byte stream native sources (especially fetch responses).
This patch does the following:
- Remove custom binding for ReadableStream and reader cancel methods. We update the binding generator so that methods can return their own promise. This is necessary to support the byte stream code path as well as the default stream code path (JS built-in).
- We do a small refactoring to simplify the construction of a ReadableStreamDefaultReader so that it always has a ReadableStream. This is now necessary for the cancel code path.
- We update getReader accordingly, which gets closer to the step-by-step spec.
- We make ReadableStream a ContextDestructionObserver so that we can get its JSDOMGlobalObject, which is used when cancelling with a DOMException (used by native sources).

We also update the conversion of DOMPromise to JSValue by using guardedObject() directly.
This removes hitting the ASSERT in DOMPromise::promise(), which is similar to the the regular JSDOMPromiseDeferred.h callPromiseFunction code path.

Covered by existing tests.

Canonical link: https://commits.webkit.org/301120@main
